### ISSUE TITLE
Indenting arguments

### ISSUE THREAD

**Issue by SimonDanisch** [jrevels/YASGuide#16]
I agree with all points, but this one:

>If an argument list extends to multiple lines, align new lines to the parenthesis/bracket/etc. that started the argument list.

I really hate this format:
```Julia
function longfunctionname_that_is_long(lots, of, args,
                                       even, more, args)
    ..code..
end
```
It's for the following reasons:

1) it makes the space that is left for arguments per line depend on the length of the function name.
2) I personally find it hard to read (I'm guessing this is highly subjective, but the irregularity of where the remaining arguments starts makes me jump back and forth while reading)
3) I still haven't found a way to let my editor help me with aligning this, so I manually need to insert spaces until it fits, which I find highly tedious and awkward to work with when I maintain/change the function in the future.
4) there isn't as much room to easily add new typing to the arguments, and it pushes you to go for shorter argument names, since there is less space.

I'd much rather prefer this format:
https://github.com/domluna/JuliaFormatter.jl/issues/9
```julia
function longfunctionname_that_is_long(
        lots, 
        of, 
        args,
        even, 
        more, 
        args
    )
    ..code..
end
```
It has a nice uniform appearance, it's very easy to see each individual argument and parse it, and it's easier to maintain the list of arguments.

One can also use JuliaFormatter.jl to migrate to this format, and also use it to enforce the YASGuide

**Comment by ararslan** [comment]
I actually find this less readable due to the placement of the closing parentheses. FWIW this rule of the YASGuide is what makes me prefer it over [Blue](https://github.com/invenia/BlueStyle), which does do what you're suggesting.

**Comment by SimonDanisch** [comment]
That's why I listed more arguments than readability, because readability is highly subjective ;)

I for once, like the closing paretheses, since they make a clear "here I end, lets put some space between the next stuff that happens" statement :)

**Comment by SimonDanisch** [comment]
Of course one could argue, that all my arguments are somewhat subjective^^

I think @jrevels mentioned, that it would be convenient to drop the YASGuide in favor of the Blue one, to have something more up to date, and prefer whatever JuliaFormatter.jl does ;)

**Comment by ararslan** [comment]
That would be incredibly unfortunate.

**Comment by SimonDanisch** [comment]
Incredible unfortunate seems like an incredible strong statement :D 
I'd be willing to change my style to conform, despite the fact that I've put lots of time and consideration into it...
You could start convincing me, by teaching me how to format to:
```julia
function longfunctionname_that_is_long(lots, of, args,
                                       even, more, args)
```
Without manually inputting lots of spaces, because you start a fight against the IDE instead of working with it...

**Comment by ararslan** [comment]
> because you start a fight against the IDE instead of working with it...

That depends on the editor. I use Vim, which you have to fight to do Blue-style indentation; it does YAS-style by default.

**Comment by SimonDanisch** [comment]
I see ;) So I guess you can understand how I feel :D

**Comment by ararslan** [comment]
What editor do you use? Surely there's a setting for it.

**Comment by SimonDanisch** [comment]
Atom + Juno! Is there one for Vim?

**Comment by jrevels** [comment]
> I think @jrevels mentioned, that it would be convenient to drop the YASGuide in favor of the Blue one, to have something more up to date

Just for clarity: I do not want to drop YASGuide in favor of the Blue Style Guide. I've stated that I'd be amenable to merging the two guides, and would consider deprecating the YASGuide in favor of the merged guide. That's my preferred outcome, actually, but I don't have time to do the merge work myself. If somebody was to make an effort at a first pass at that, though, I'd review it. It'll probably take a lot of bikeshedding, though 😄 

> prefer whatever JuliaFormatter.jl does

It's not that I would prefer JuliaFormatter's enforced style, it's just that I think it's pretty hard to enforce style without an auto-formatter, so at the end of the day whichever style guide has an accompanying auto-formatter will probably end up winning over the most users.

**Comment by ararslan** [comment]
> Is there one for Vim?

Not that I found. Jarrett also uses Atom, so perhaps he can provide some guidance.

**Comment by domluna** [comment]
I use vim and am sympathetic to @ararslan but once you get things in the form

```
func_a(
    arg1,
)
```

it works well. It's just initially vim opens it at `(`. I'm sure there's a way to override this though.

**Comment by jrevels** [comment]
(It's a Friday night, perfect time to paint a bikeshed!)

It's mostly subjective, but the main things I don't like about the proposed format are:

- The amount of space allocated to each argument is usually going to be way more than feels merited. The drastic drop in density causes my brain to unnecessarily downshift when skimming, as the arguments jarringly
    pop, 
    out, 
    as, 
    if,
    each,
    line,
    expresses more information/importance than each line actually does. 
- The closing bracket on its own newline is noisy to me basically for the same reason.
- The result doesn't look like a natural continuation of the short-form style.

The main downsides of the current format seem to be folks having alignment and line length issues. 

I'm not too concerned about the alignment issue, I guess because I don't feel the same pain (maybe I'm just used to it?) but also because I care way less about writability than readability and I feel like this is primarily an issue with the former. 

The line length issue, though - I sympathize with that.

I had the following idea that mostly keeps the properties I find nice about the current rules while solving the line length issue. I'm still kinda leaning against it, but maybe it'll grow on me overnight:

```julia
function long_name_of_function_because_i_am_writing_an_example(
        arg1, arg2, arg3, arg4,
        arg5, arg6, arg7, arg8)
    my_code = starts_here()
```

So basically, you're allowed to pop a single newline after the opening paren, but no newline before the closing paren and you're allowed to have multiple arguments per line. The double-indent on the argument list keeps the argument list visually separated from the single-indented code block without being as noisy as a closing separator. 

My main concern by allowing this is that people would "abuse" it for function names that aren't actually long enough for this to matter...

**Comment by ararslan** [comment]
That's somewhat reasonable but I feel like that introduces inconsistency where none currently exists.

**Comment by domluna** [comment]
I've been thinking about initially just moving the args to their own line and see if that fits, i.e.

```julia
function long_name_of_function_because_i_am_writing_an_example(
    arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8
)
    my_code = starts_here()
```

then if that doesn't fit:

```julia
function long_name_of_function_because_i_am_writing_an_example(
    arg1,
    arg2,
    arg3,
    arg4,
    arg5,
    arg6,
    arg7,
    arg8,
)
    my_code = starts_here()
```

> The closing bracket on its own newline is noisy to me basically for the same reason

What's a bracket gotta do around here to get its own newline? 

Pretty much all formatters do this. In my opinion the main benefit is acting as a clear divisor between code blocks, in this case, the function arguments and the function body. It also makes it easier to add a new argument.

**Comment by jrevels** [comment]
I've been trying to keep [my proposed rule](https://github.com/jrevels/YASGuide/issues/16#issuecomment-554588796) in my head while reading code recently, just to see where it might come up in practice. My conclusion: situations where this exceptional rule would be needed are rare and otherwise avoidable anyway, so I'm going to close this for now.

**Comment by Sacha0** [comment]
FWIW, I favor the approach in https://github.com/jrevels/YASGuide/issues/16#issuecomment-554588796 or similar due to not irregularly encountering definitions like the following in the wild:
```
function long_name_of_function_like_those_above{MaybeConstructor}(long_fluffy_variable_1{A},
                                                                  long_fluffy_variable_2{B},
                                                                  long_fluffy_variable_3{C},
                                                                  long_fluffy_variable_4{D},
                                                                  long_fluffy_variable_5{E},
                                                                  long_fluffy_variable_6{F},
                                                                  long_fluffy_variable_7{G},
                                                                  long_fluffy_variable_8{H}) where
                                                                  {A<:Tuple,B<:Integer,C<:Cat,
                                                                  D<:Dog,E<:Egg,F<:FluffyObject,
                                                                  G<:Gratuitous,H<:Halp}
    the_code = starts_here
    [...]
end                                 
```
which often seems nicer to be able to write as
```
function long_name_of_function_like_those_above{MaybeConstructor}(
        long_fluffy_variable_1{A}, long_fluffy_variable_2{B}, long_fluffy_variable_3{C},
        long_fluffy_variable_4{D}, long_fluffy_variable_5{E}, long_fluffy_variable_6{F},
        long_fluffy_variable_7{G}, long_fluffy_variable_8{H}) where
        {A<:Tuple,B<:Integer,C<:Cat,D<:Dog,E<:Egg,F<:FluffyObject,G<:Gratuitous,H<:Halp}
    the_code = starts_here
    [...]
end
```
or
```
function long_name_of_function_like_those_above{MaybeConstructor}(long_fluffy_variable_1{A},
        long_fluffy_variable_2{B}, long_fluffy_variable_3{C}, long_fluffy_variable_4{D},
        long_fluffy_variable_5{E}, long_fluffy_variable_6{F}, long_fluffy_variable_7{G},
        long_fluffy_variable_8{H}) where
        {A<:Tuple,B<:Integer,C<:Cat,D<:Dog,E<:Egg,F<:FluffyObject,G<:Gratuitous,H<:Halp}
    the_code = starts_here
    [...]
end
```
Perhaps there's still room something like one or the other in YASGuide? :)

**Comment by jrevels** [comment]
Yeah, I can sympathize there, I run into this sometimes too. I usually then refactor the code in other ways to avoid the ugliness lol (similar to https://github.com/domluna/JuliaFormatter.jl/issues/263#issuecomment-711187061). That generally results in better-written code overall, actually, but perhaps a high bar to leap over for e.g. an autoformatter 😁 

this is the kind of thing that I'd be more inclined to consider if there was a JuliaFormatter patch that formatted code this way so that I could see what it looks like on real code

**Comment by Sacha0** [comment]
Agreed, refactoring tends to be best in that situation, though regrettably the necessary refactoring is often an unavailable luxury in such cases 😄.
